### PR TITLE
Tpetra: document the requirements of column maps

### PR DESCRIPTION
**Pull request by brian-kelley**
At CrsMatrix/CrsGraph constructors and replaceColMap(),
document the GID ordering requirements of column maps:
-locals before remotes (determined by domain map)
-locals sorted
-remotes sorted first by proc and then by GID

Strongly recommending that Details::makeColMap() be
used instead of constructing it manually from raw GIDs.

Also, fix a typo in a makeColMap comment.

@trilinos/tpetra 

## Motivation

The requirements of column maps need to be codified, especially as more matrix/graph construction stuff gets rewritten using Kokkos. Having these comments here would have made #6190 a lot easier, for example.

## Related Issues

* Closes 
* Blocks 
* Is blocked by 
* Follows #6190  
* Precedes 
* Related to 
* Part of 
* Composed of 

## Testing

Just documentation (doxygen/comment) changes.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@brian-kelley Those are not actually requirements of CrsGraph or CrsMatrix.  Both of those classes work just fine with arbitrary column Maps.  We've exercised those cases in Nalu, for example.  The issue is that certain preconditioners have those requirements.

File: packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp [hunk 1]
```diff
@@ -411,6 +411,7 @@ namespace Tpetra {
     /// \param rowMap [in] Distribution of rows of the graph.
     ///
     /// \param colMap [in] Distribution of columns of the graph.
+    ///   See replaceColMap() for the requirements.
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp]: This is not correct.  CrsGraph itself can work with any column Map.

File: packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp [hunk 2]
```diff
@@ -1540,6 +1546,16 @@ namespace Tpetra {
     /// does their order change, as a result of calling this method.
     ///
     /// \param newColMap [in] New column Map.  Must be nonnull.
+    ///   The column map's list of global indices on each process must be in a certain order:
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp]: This is not correct.  CrsGraph itself can work with any column Map.

**Comment by brian-kelley** [comment]
@mhoemmen OK, maybe this isn't the right place to put these comments/requirements. Basically, if somebody comes along later to do what I tried to do with the add (which is to build the column map from scratch for efficiency), they should know what the de facto requirements are for stuff to work downstream. It took several tries for me to get #6190 to work because I didn't understand what was causing MueLu/Xpetra/Ifpack2 tests to fail. So I would still like to codify them somewhere that's easy to find.

@jhux2 @csiefer2 @cgcgcg @lucbv Do you have any thoughts on this, about where column map ordering requirements could be documented?

**Comment by mhoemmen** [comment]
@brian-kelley I think it's fine to put the requirements in CrsGraph's documentation, but they are not CrsGraph requirements.  We could say something like "some downstream packages, including preconditioners in Ifpack2 and MueLu, have the following requirements on column Maps ...."

**Comment by brian-kelley** [comment]
@mhoemmen OK, thanks. I just made those changes. Could @trilinos/muelu weigh in? Are the new comments correct, and can we make the restrictions more specific/less conservative?

**Comment by jhux2** [comment]
> I think it's fine to put the requirements in CrsGraph's documentation, but they are not CrsGraph requirements. We could say something like "some downstream packages, including preconditioners in Ifpack2 and MueLu, have the following requirements on column Maps ...."

This could go in the Ifpack2 and MueLu user guides and/or doxygen for the preconditioner adapters.

File: packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp [hunk 3]
```diff
@@ -1546,11 +1546,13 @@ namespace Tpetra {
     /// does their order change, as a result of calling this method.
     ///
     /// \param newColMap [in] New column Map.  Must be nonnull.
-    ///   The column map's list of global indices on each process must be in a certain order:
+    ///   Within Tpetra, there are no particular restrictions on the column map.
+    ///   However, if this graph will be used in Xpetra, Ifpack2, or MueLu,
+    ///   the column map's list of global indices on each process must be in a certain order:
     ///   locally owned GIDs (sorted), followed by remote GIDs (in order of
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp]: Locally owned GIDs don't have to be sorted; they just need to be in the same order as in the domain Map.  The domain Map defines the "locally owned GIDs" for each process.

File: packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp [hunk 4]
```diff
@@ -1546,11 +1546,13 @@ namespace Tpetra {
...
     ///   owning proc, and sorted within each proc).
     ///
-    ///   It is strongly recommended to use Tpetra::Details::makeColMap()
+    ///   If in doubt, it is strongly recommended to use Tpetra::Details::makeColMap()
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp]: I think it's fair to say "It is strongly recommended...."

File: packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp [hunk 5]
```diff
@@ -2389,11 +2389,13 @@ namespace Tpetra {
     /// \brief Replace the matrix's column Map with the given Map.
     ///
     /// \param newColMap [in] New column Map.  Must be nonnull.
-    ///   The column map's list of global indices on each process must be in a certain order:
+    ///   Within Tpetra, there are no particular restrictions on the column map.
+    ///   However, if this matrix will be used in Xpetra, Ifpack2, or MueLu,
+    ///   the column map's list of global indices on each process must be in a certain order:
     ///   locally owned GIDs (sorted), followed by remote GIDs (in order of
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp]: See above.

File: packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp [hunk 6]
```diff
@@ -2389,11 +2389,13 @@ namespace Tpetra {
...
     ///   owning proc, and sorted within each proc).
     ///
-    ///   It is strongly recommended to use Tpetra::Details::makeColMap()
+    ///   If in doubt, it is strongly recommended to use Tpetra::Details::makeColMap()
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp]: See above.

**Comment by jhux2** [comment]
@brian-kelley wrote:
> OK, thanks. I just made those changes. Could @trilinos/muelu weigh in? Are the new comments correct, and can we make the restrictions more specific/less conservative?

These comments looks good to me.

**Comment by mhoemmen** [comment]
Locally owned GIDs don't have to be sorted in the column Map; they just need to be in the same order as in the domain Map. The domain Map defines the "locally owned GIDs" for each process.

**Comment by jhux2** [comment]
> Locally owned GIDs don't have to be sorted in the column Map; they just need to be in the same order as in the domain Map.

@mhoemmen Good catch.

**Comment by brian-kelley** [comment]
I removed "If in doubt", and said that this is called Aztec ordering. I ran Doxygen to make sure everything looks right.

**Pull request merged**